common/messaging/kafka/consumer_impl_test.go

```diff
@@ -0,0 +1,217 @@
+	assert.Equal(t, kafkaProducer, consumerHandler.dlqProducer)
+	assert.Equal(t, topic, consumerHandler.topic)
+	// Close the channel at the end of the test
+	close(msgChan)
+}
+
+// test multiple methods related to messageImpl since setup is repeated
+func TestMessageImpl(t *testing.T) {
+	topic := "test-topic"
+	metricsClient := metrics.NewClient(tally.NoopScope, metrics.History)
+	logger := testlogger.New(t)
+	mockProducer := mocks.NewSyncProducer(t, nil)
+	kafkaProducer := NewKafkaProducer(topic, mockProducer, logger)
+	msgChan := make(chan messaging.Message, 1)
+	consumerHandler := newConsumerHandlerImpl(kafkaProducer, topic, msgChan, metricsClient, logger)
+	consumerGroupSession := NewMockConsumerGroupSession(int32(1))
+	partition := int32(100)
+	offset := int64(0)
+	msgImpl := &messageImpl{
+		saramaMsg: &sarama.ConsumerMessage{
+			Topic:     topic,
+			Partition: partition,
+			Offset:    offset,
+		},
+		session: consumerGroupSession,
+		handler: consumerHandler,
+		logger:  logger,
+	}
+
+	// Ack message that is from a previous session
+	msgImpl.handler.Setup(NewMockConsumerGroupSession(int32(2)))
+	err := msgImpl.Ack()
+	assert.NoError(t, err)
+
+	// normal case
+	msgImpl.handler.Setup(NewMockConsumerGroupSession(int32(1)))
+	msgImpl.handler.manager.AddMessage(partition, offset)
+	err = msgImpl.Ack()
+	assert.NoError(t, err)
+
+	// Nack message that is from a previous session
+	msgImpl.handler.Setup(NewMockConsumerGroupSession(int32(2)))
+	err = msgImpl.Nack()
+	assert.NoError(t, err)
+
+	// normal case
+	msgImpl.handler.Setup(NewMockConsumerGroupSession(int32(1)))
+	mockProducer.ExpectSendMessageAndSucceed()
+	msgImpl.handler.manager.AddMessage(partition, offset)
+	err = msgImpl.Nack()
+	assert.NoError(t, err)
+
+	close(msgChan)
+}
+
+// MockConsumerGroupSession implements sarama.ConsumerGroupSession for testing purposes.
+type MockConsumerGroupSession struct {
+	claims       map[string][]int32
+	memberID     string
+	generationID int32
+	offsets      map[string]map[int32]int64 // You can store offsets here to verify MarkOffset calls.
+	commitCalled bool
+	mu           sync.Mutex
+	context      context.Context
+}
+
+func NewMockConsumerGroupSession(generationID int32) *MockConsumerGroupSession {
+	return &MockConsumerGroupSession{
+		claims:       map[string][]int32{},
+		memberID:     "test-member",
+		generationID: generationID,
+		offsets:      map[string]map[int32]int64{},
+		commitCalled: false,
+	}
+}
+
+func (m *MockConsumerGroupSession) Claims() map[string][]int32 {
+	return m.claims
+}
+
+func (m *MockConsumerGroupSession) MemberID() string {
+	return m.memberID
+}
+
+func (m *MockConsumerGroupSession) GenerationID() int32 {
+	return m.generationID
+}
+
+func (m *MockConsumerGroupSession) MarkOffset(topic string, partition int32, offset int64, metadata string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.offsets[topic] == nil {
+		m.offsets[topic] = make(map[int32]int64)
+	}
+	m.offsets[topic][partition] = offset
```

Comment:
did you mean to validate final value of `offsets` in the test?

Reply:
It is part of the interface. When complete the message, it will mark the provided offset, alongside a metadata string that represents the state of the partition consumer at that point in time. The test didn't need this, I will just make it empty.